_posts/2020-12-09-#20477.md

```diff
@@ -10,13 +10,29 @@ status: upcoming
 commit:
 ---
 
-_Notes and questions to follow soon!_
-
-<!-- TODO: Before meeting, add notes and questions
 ## Notes
+* When all inbound connection slots are occupied and a new inbound connection is received, the node attempts to evict an existing inbound connection in order to open up a slot.
+* However, any eviction logic is susceptible to [eclipse attacks](https://www.avivz.net/pubs/15/sec15-paper-heilman_final.pdf). An attacker may try to make lots of new inbound connections to us, and if we always evict another honest node, we will eventually be eclipsed.
+* `AttemptToEvictConnection` in `src/net.cpp` contains logic to increase the cost of such attacks. Among other things, in order to not get their own previous connections evicted when they make new inbounds, the attacker would need to:
+	- Obtain IP addresses in various net groups
+	- Be amongst the fastest of our peers (geographic cost)
+	- Provide us novel transactions risking fees
+	- Provide us with novel blocks
+* The scarcity of tests for the eviction logic was first noticed in [#16660](https://github.com/bitcoin/bitcoin/issues/16660) and addressed with functional tests in [#16756](https://github.com/bitcoin/bitcoin/pull/16756) ([review club](https://bitcoincore.reviews/16756)).
+* This PR, [#20477](https://github.com/bitcoin/bitcoin/pull/20477) refactors some of `AttemptToEvictConnection` into `SelectNodeToEvict` to increase testability and adds unit tests for `SelectNodeToEvict`.
 
 ## Questions
--->
+1. Why is it important to allow for evictions? What are the risks in rejecting new inbounds if slots are full?
+2. What are the various ways `AttemptToEvictConnection` tries to increase the cost of an eclipse attack?
+3. Why does `AttemptToEvictConnection` [remove peers being disconnected from eviction candidates](https://github.com/bitcoin/bitcoin/blob/master/src/net.cpp#L932) instead of waiting for them to disconnect (as that would open up a slot)?
+4. What is a net group? Why does Bitcoin Core group peers by [ASNs](https://hackertarget.com/as-ip-lookup/) instead of /16 subnets [since](https://github.com/bitcoin/bitcoin/issues/16599) [0.20](https://github.com/bitcoin/bitcoin/pull/16702)?
+5. Can an attacker predict which net groups are most likely to offer them protection from eviction?
+6. Why is eviction guaranteed if we have at least 29 eviction candidates?
+7. Why is non-eviction guaranteed if we have no more than 20 eviction candidates? Is 20 the highest number of nodes that guarantees non-eviction?
```

Comment:
I recall you were asking about this number; test/functional/p2p_eviction.py::L40-41 contains some notes about it, if helpful.

Reply:
That's a helpful comment. I think I now understand. Do you agree that the protection at net.cpp::961 for up to 8 non-tx-relay peers may, or may not apply to the randomly generated eviction candidates since `!n.fRelayTxes && n.fRelevantServices` will evaluate to `randbool`. So we cannot count them in the _guaranteed_ eviction?